homeassistant/components/tplink/camera.py

```diff
@@ -0,0 +1,218 @@
+        """Return unique ID for the entity."""
+        return f"{legacy_device_id(self._device)}-{self.entity_description}"
+
+    @callback
+    def _async_update_attrs(self) -> None:
+        """Update the entity's attributes."""
+        self._attr_is_on = self._camera_module.is_on
+
+    async def stream_source(self) -> str | None:
+        """Return the source of the stream."""
+        return self._video_url
+
+    async def _async_check_stream_auth(self, video_url: str) -> None:
+        """Check for an auth error and start reauth flow."""
+        if await async_has_stream_auth_error(self.hass, video_url):
+            _LOGGER.debug(
+                "Camera stream failed authentication for %s",
+                self._device.host,
+            )
+            self._can_stream = False
+            self.coordinator.config_entry.async_start_reauth(
+                self.hass,
+                ConfigFlowContext(
+                    reauth_source=CONF_CAMERA_CREDENTIALS,  # type: ignore[typeddict-unknown-key]
+                ),
+                {"device": self._device},
+            )
+
+    async def async_camera_image(
+        self, width: int | None = None, height: int | None = None
+    ) -> bytes | None:
+        """Return a still image response from the camera."""
+        now = dt_util.utcnow()
+
+        if self._image and now - self._last_update < self.IMAGE_INTERVAL:
+            return self._image
+
+        # Don't try to capture a new image if a stream is running
+        if (self.stream and self.stream.available) or self._http_mpeg_stream_running:
+            return self._image
+
+        if self._can_stream and (video_url := self._video_url):
+            # Sometimes the front end makes multiple image requests
+            async with self._image_lock:
+                if self._image and now - self._last_update < self.IMAGE_INTERVAL:
+                    return self._image
+
+                _LOGGER.debug("Updating camera image for %s", self._device.host)
+                image = await ffmpeg.async_get_image(
+                    self.hass,
+                    video_url,
+                    width=width,
+                    height=height,
+                )
+                if image:
+                    self._image = image
+                    self._last_update = now
+                    _LOGGER.debug("Updated camera image for %s", self._device.host)
+                # This coroutine is called by camera with an asyncio.timeout
+                # so image could be None whereas an auth issue returns b''
+                elif image == b"":
+                    _LOGGER.debug(
+                        "Empty camera image returned for %s", self._device.host
+                    )
+                    # image could be empty if a stream is running so check for explicit auth error
+                    await self._async_check_stream_auth(video_url)
+                else:
+                    _LOGGER.debug(
```

Comment:
Looks like coverage is missing on this line

Reply:
Fixed